dsp/window/window_parametric.go

```diff
@@ -0,0 +1,165 @@
+	}
+	return seq
+}
+
+// Tukey can modify a sequence using the Tukey window and return the result.
+// See https://en.wikipedia.org/wiki/Window_function#Tukey_window
+// and https://www.recordingblogs.com/wiki/tukey-window for details.
+//
+// The Tukey window is an adjustable window.
+//
+// The sequence weights are
+//  w[k] = 0.5 * (1 + cos(π*(|k - M| - αM)/((1-α) * M))), |k - M| ≥ αM
+//       = 1, |k - M| < αM
+// with M = (N - 1)/2 for k=0,1,...,N-1 where N is the length of the window.
+//
+// Spectral leakage parameters are summarized in the table:
+//         |  α=0.3 |  α=0.5 |  α=0.7 |
+//  -------|--------------------------|
+//  ΔF_0   |   1.33 |   1.22 |   1.13 |
+//  ΔF_0.5 |   1.28 |   1.16 |   1.04 |
+//  K      |   0.67 |   0.61 |   0.57 |
+//  ɣ_max  | -18.2  | -15.1  | -13.8  |
+//  β      |  -1.41 |  -2.50 |  -3.74 |
+type Tukey struct {
+	Alpha float64
+}
+
+// Transform applies the Tukey transformation to seq in place, using the value
+// of the receiver as the Alpha parameter, and returning the result
+func (t Tukey) Transform(seq []float64) []float64 {
+	switch {
+	case t.Alpha <= 0:
+		return Rectangular(seq)
+	case t.Alpha >= 1:
+		return Hann(seq)
+	default:
+		alphaL := t.Alpha * float64(len(seq)-1)
+		width := int(0.5*alphaL) + 1
+		for i := range seq[:width] {
+			w := 0.5 * (1 - math.Cos(2*math.Pi*float64(i)/alphaL))
+			seq[i] *= w
+			seq[len(seq)-1-i] *= w
+		}
+		return seq
+	}
+}
+
+// TransformComplex applies the Tukey transformation to seq in place, using the value
+// of the receiver as the Alpha parameter, and returning the result
+func (t Tukey) TransformComplex(seq []complex128) []complex128 {
+	switch {
+	case t.Alpha <= 0:
+		return RectangularComplex(seq)
+	case t.Alpha >= 1:
+		return HannComplex(seq)
+	default:
+		alphaL := t.Alpha * float64(len(seq)-1)
+		width := int(0.5*alphaL) + 1
+		for i, v := range seq[:width] {
+			w := 0.5 * (1 - math.Cos(2*math.Pi*float64(i)/alphaL))
+			v = complex(w*real(v), w*imag(v))
+			seq[i] = v
+			seq[len(seq)-1-i] = v
+		}
+		return seq
+	}
+}
+
+// Values is an arbitrary real window function.
+type Values []float64
+
+// NewValues returns a Values of length n with weights corresponding to the
+// provided window function.
+func NewValues(window func([]float64) []float64, n int) Values {
+	v := make(Values, n)
+	for i := range v {
+		v[i] = 1
+	}
+	return window(v)
+}
+
+// Transform applies the weights in the receiver to seq in place, returning the
+// result. If v is nil, Transform is a no-op, otherwise the length of v must
+// match the length of seq.
+func (v Values) Transform(seq []float64) []float64 {
+	if v == nil {
+		return seq
+	}
+	if len(v) != len(seq) {
+		panic("window: length mismatch")
+	}
+	for i, w := range v {
+		seq[i] *= w
+	}
+	return seq
+}
+
+// TransformComplex applies the weights in the receiver to seq in place, returning
+// the result. If v is nil, Transform is a no-op, otherwise the length of v must
```

Comment:
TransformComplex also here. Sorry I didn't see it before.

Reply:
... or I could use some initiative.